instadam/project.py

```diff
@@ -225,3 +225,79 @@ def get_labels(project_id):
     labels = [{'name': label.label_name, 'id': label.id} for label in
               project.labels]
     return jsonify({'labels': labels}), 200
+
+
+@bp.route('/project/<project_id>/permissions', methods=['PUT'])
+@jwt_required
+def update_user_permission(project_id):
+    """ Grant a user with specified privilege to project
+
+        Grant a user with specified privilege (READ_WRITE or READ_ONLY) to project
+        upon receiving a `PUT`request to the `/project/<project_id>/permissions`
+        entry point. User must be signed in as an ADMIN and have READ_WRITE
+        permission to the project. User must provide a `username` to specify the
+        user that will be granted permissions, and `access_type` to specify the
+        type of privilege to grant.
+
+        `access_type` takes a string of two values: `r` or `rw`, where `r` is
+        `READ_ONLY` and `rw` is `READ_WRITE`
+
+        Must supply a jwt-token to verify user status and extract `user_id`.
+
+        Raises:
+            400 Error if user_name or access_type is not specified
+            400 Error if access_type is not 'r' or 'rw'
+            401 Error if not logged in
+            401 Error if user is not an ADMIN
+            401 Error if user does not have privilege to update permission
+            404 Error if user with user_name does not exist
+
+        Returns:
+            201 if success. Will also return `project_id`.
+        """
+
+    map_code_to_access_type = {
+        'r': AccessTypeEnum.READ_ONLY,
+        'rw': AccessTypeEnum.READ_WRITE,
+    }
+
+    project = maybe_get_project(project_id)  # check privilege and get project
+
+    req = request.get_json()
+    if 'username' not in req:
+        abort(400, 'user_name must be included.')
+    if 'access_type' not in req:
+        abort(400, 'access_type must be included.')
+
+    username = req['username']
+    user = User.query.filter_by(username=username).first()
+    if user is None:
+        abort(404, 'User with username=%s does not exist' % username)
+
+    if req['access_type'] not in map_code_to_access_type:
+        abort(400, 'Not able to interpret access_type.')
+
+    access_type = map_code_to_access_type[req['access_type']]
+
+    # Check if user already have the permission of `access_type` to the project
+    permission = ProjectPermission.query.filter(
```

Comment:
Maybe check whether the requested user is admin, since only admin can have rw access to a project. If the requested user is annotator, he/she can only have r access.

Reply:
Updated